internal/transformers/estransforms/optionalchain.go

```diff
@@ -11,7 +11,222 @@ type optionalChainTransformer struct {
+		return ch.visitNonOptionalExpression(node.Expression, false, true)
+	}
+	return ch.Visitor().VisitEachChild(node.AsNode())
+}
+
+func (ch *optionalChainTransformer) visitNonOptionalExpression(node *ast.Expression, captureThisArg bool, isDelete bool) *ast.Expression {
+	switch node.Kind {
+	case ast.KindParenthesizedExpression:
+		return ch.visitParenthesizedExpression(node.AsParenthesizedExpression(), captureThisArg, isDelete)
+	case ast.KindElementAccessExpression, ast.KindPropertyAccessExpression:
+		return ch.visitPropertyOrElementAccessExpression(node, captureThisArg, isDelete)
+	case ast.KindCallExpression:
+		return ch.visitCallExpression(node.AsCallExpression(), captureThisArg)
+	default:
+		return ch.Visitor().VisitNode(node.AsNode())
+	}
+}
+
+type flattenResult struct {
+	expression *ast.Expression
+	chain      []*ast.Node
+}
+
+func flattenChain(chain *ast.Node) flattenResult {
+	// Debug.assertNotNode(chain, isNonNullChain); // !!!
+	links := []*ast.Node{chain}
+	for !ast.IsTaggedTemplateExpression(chain) && chain.QuestionDotToken() == nil {
+		chain = ast.SkipPartiallyEmittedExpressions(chain.Expression())
+		// Debug.assertNotNode(chain, isNonNullChain); // !!!
+		links = append([]*ast.Node{chain}, links...)
+	}
+	return flattenResult{chain.Expression(), links}
+}
+
+func isCallChain(node *ast.Node) bool {
+	return ast.IsCallExpression(node) && node.Flags&ast.NodeFlagsOptionalChain != 0
+}
+
+func (ch *optionalChainTransformer) visitOptionalExpression(node *ast.Node, captureThisArg bool, isDelete bool) *ast.Node {
+	r := flattenChain(node)
+	expression := r.expression
+	chain := r.chain
+	left := ch.visitNonOptionalExpression(ast.SkipPartiallyEmittedExpressions(expression), isCallChain(chain[0]), false)
+	var leftThisArg *ast.Expression
+	capturedLeft := left
+	if ast.IsSyntheticReferenceExpression(left) {
+		leftThisArg = left.AsSyntheticReferenceExpression().ThisArg
+		capturedLeft = left.AsSyntheticReferenceExpression().Expression
+	}
+	leftExpression := ch.Factory().RestoreOuterExpressions(expression, capturedLeft, ast.OEKPartiallyEmittedExpressions)
+	if !transformers.IsSimpleCopiableExpression(capturedLeft) {
+		capturedLeft = ch.Factory().NewTempVariable()
+		ch.EmitContext().AddVariableDeclaration(capturedLeft)
+		leftExpression = ch.Factory().NewAssignmentExpression(capturedLeft, leftExpression)
+	}
+	rightExpression := capturedLeft
+	var thisArg *ast.Expression
+
+	for i, segment := range chain {
+		switch segment.Kind {
+		case ast.KindElementAccessExpression, ast.KindPropertyAccessExpression:
+			if i == len(chain)-1 && captureThisArg {
+				if !transformers.IsSimpleCopiableExpression(rightExpression) {
+					thisArg = ch.Factory().NewTempVariable()
+					ch.EmitContext().AddVariableDeclaration(thisArg)
+					rightExpression = ch.Factory().NewAssignmentExpression(thisArg, rightExpression)
+				} else {
+					thisArg = rightExpression
+				}
+			}
+			if segment.Kind == ast.KindElementAccessExpression {
```

Comment:
nit - the check is inverted compared to the original code

Reply:
As are the bodies in the branches of the check. `E` comes before `P`, so `E` comes first in the `case` clause list, so handling the first case explicitly is a bit more natural looking.